onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -1957,4 +1925,60 @@ CUDAExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
   return result;
 }
 
+void CUDAExecutionProvider::RegisterAllocator(std::shared_ptr<AllocatorManager> allocator_manager) {
+  // Try to get a CUDA allocator from allocator manager first
+  // Used to allocate CUDA device memory
+  auto cuda_alloc = allocator_manager->GetAllocator(info_.device_id, OrtMemTypeDefault);
+  if (nullptr == cuda_alloc) {
+    AllocatorCreationInfo default_memory_info(
+        [](OrtDevice::DeviceId device_id) {
+          return onnxruntime::make_unique<CUDAAllocator>(device_id, CUDA);
+        },
+        info_.device_id,
+        true,
+        {info_.cuda_mem_limit,
+         static_cast<int>(info_.arena_extend_strategy),
+         -1, -1});
+    cuda_alloc = CreateAllocator(default_memory_info);
+    allocator_manager->InsertAllocator(cuda_alloc);
+  }
+  InsertAllocator(cuda_alloc);
```

Comment:
If the EP has access to the allocator_mgr instance, does it still need to store the allocator in its own instance? It can potentially lead to inconsistencies.

Reply:
The finally goal is to keep all allocators in the central place allocator_mgr instance. It requires lots of changes which will impact all EPs. Like we discussed las week, let's get the approach agreed first, then we can move further. I will have another follow up PR to achieve this. then CPUAllocator can be shared across all EPs.

---
In reply to: [563003537](https://github.com/microsoft/onnxruntime/pull/6332#discussion_r563003537) [](ancestors = 563003537)